File: qviability/export/excel.py
"""Excel export — one row per image, all measurement metrics as columns."""

from __future__ import annotations

import pandas as pd

from ..core.project import Project

COLUMNS = [
    "Filename", "Channel", "Positive pixels", "% above threshold",
]


def build_dataframe(project: Project) -> pd.DataFrame:
    """One row per image, sorted by channel then filename."""
    rows = []
    for e in project.images:
        m = e.measurement
        rows.append({
            "Filename": e.display_name,
            "Channel": e.channel,
            "Positive pixels": m.positive_pixels if m else None,
            "% above threshold": round(m.positive_area_pct, 4) if m else None,
        })
    df = pd.DataFrame(rows, columns=COLUMNS)
    if not df.empty:
        df = df.sort_values(["Channel", "Filename"]).reset_index(drop=True)
    return df


def write_xlsx(project: Project, path: str) -> None:
    """Write the results workbook to *path*."""
    build_dataframe(project).to_excel(path, index=False, engine="openpyxl")

File: web/src/export/excel.ts
// Results workbook — port of qviability/export/excel.py.
//
// One row per image: the four desktop columns, sorted by channel then natural
// filename. Built with exceljs and returned as bytes for FSA write or ZIP.

import ExcelJS from "exceljs";
import type { Project } from "@/core/types";
import { compareNatural } from "@/core/imageIo";

export const COLUMNS = [
  "Filename",
  "Channel",
  "Positive pixels",
  "% above threshold",
] as const;

interface Row {
  filename: string;
  channel: string;
  positivePixels: number | null;
  pctAbove: number | null;
}

function rowsFor(project: Project): Row[] {
  const rows: Row[] = project.images.map((e) => {
    const m = e.measurement;
    return {
      filename: e.display_name,
      channel: e.channel,
      positivePixels: m ? m.positive_pixels : null,
      // round(x, 4) — matches pandas round on positive_area_pct.
      pctAbove: m ? Math.round(m.positive_area_pct * 1e4) / 1e4 : null,
    };
  });
  rows.sort((a, b) => {
    if (a.channel !== b.channel) return a.channel < b.channel ? -1 : 1;
    return compareNatural(a.filename, b.filename);
  });
  return rows;
}

export async function buildXlsx(project: Project): Promise<Uint8Array> {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet("Sheet1");
  ws.addRow([...COLUMNS]);
  for (const r of rowsFor(project)) {
    ws.addRow([r.filename, r.channel, r.positivePixels, r.pctAbove]);
  }
  const buf = await wb.xlsx.writeBuffer();
  return new Uint8Array(buf as ArrayBuffer);
}
